### PR TITLE
Add slog

### PR DESCRIPTION
**Pull request by dongrie**
* Add slog
* Add time track

```
{"time":"2023-09-14T10:25:00.919158+09:00","level":"INFO","source":{"function":"github.com/hyperledger-labs/yui-relayer/log.(*RelayLogger).TimeTrack","file":"/Users/dongri.jin/go/src/github.com/dongrie/yui-relayer-time-track/log/slog.go","line":182},"msg":"time track","module":"ethereum.chain","name":"findRecvPacketEvents","elapsed":10582031}
{"time":"2023-09-14T10:25:00.921196+09:00","level":"INFO","source":{"function":"github.com/hyperledger-labs/yui-relayer/log.(*RelayLogger).TimeTrack","file":"/Users/dongri.jin/go/src/github.com/dongrie/yui-relayer-time-track/log/slog.go","line":182},"msg":"time track","module":"ethereum.chain","name":"findRecvPacketEvents","elapsed":12625927}
{"time":"2023-09-14T10:25:00.921245+09:00","level":"INFO","source":{"function":"github.com/hyperledger-labs/yui-relayer/log.(*RelayLogger).TimeTrack","file":"/Users/dongri.jin/go/src/github.com/dongrie/yui-relayer-time-track/log/slog.go","line":182},"msg":"time track","module":"ethereum.chain","name":"findReceivedPackets","elapsed":12800764}
{"time":"2023-09-14T10:25:00.927123+09:00","level":"INFO","source":{"function":"github.com/hyperledger-labs/yui-relayer/log.(*RelayLogger).TimeTrack","file":"/Users/dongri.jin/go/src/github.com/dongrie/yui-relayer-time-track/log/slog.go","line":182},"msg":"time track","module":"ethereum.chain","name":"findWriteAckEvents","elapsed":6640852}
{"time":"2023-09-14T10:25:00.927168+09:00","level":"INFO","source":{"function":"github.com/hyperledger-labs/yui-relayer/log.(*RelayLogger).TimeTrack","file":"/Users/dongri.jin/go/src/github.com/dongrie/yui-relayer-time-track/log/slog.go","line":182},"msg":"time track","module":"ethereum.chain","name":"findReceivedPackets","elapsed":18725720}
{"time":"2023-09-14T10:25:00.934499+09:00","level":"INFO","source":{"function":"github.com/hyperledger-labs/yui-relayer/log.(*RelayLogger).TimeTrack","file":"/Users/dongri.jin/go/src/github.com/dongrie/yui-relayer-time-track/log/slog.go","line":182},"msg":"time track","module":"ethereum.chain","name":"QueryUnreceivedAcknowledgements","elapsed":356}
```

### CODE REVIEW

File: pkg/relay/ethereum/module.go [hunk 1]
```diff
@@ -24,3 +27,8 @@ func (Module) RegisterInterfaces(registry codectypes.InterfaceRegistry) {
 func (Module) GetCmd(ctx *config.Context) *cobra.Command {
 	return nil
 }
+
+func GetModuleLogger() *log.RelayLogger {
+	return log.GetLogger().
+		WithModule(LOGGER_MODULE)
+}
```
Comment (siburu) [inline pkg/relay/ethereum/module.go:34]: You are using this module logger everywhere in this module, but this logger contains no information about clients or connecitons or channels or anything else...?
Readers of logs can't even know which chain (src or dst) is processed, can they?

Basically we want to use `core.GetXxxLogger` also here.
If `core.GetXxxLogger` is not suitable for modules outside of core, please fix them.

File: pkg/relay/ethereum/module.go [hunk 2]
```diff
@@ -3,13 +3,16 @@ package ethereum
...
 	"github.com/hyperledger-labs/yui-relayer/config"
+	"github.com/hyperledger-labs/yui-relayer/log"
 	"github.com/spf13/cobra"
 )
 
 type Module struct{}
 
 var _ config.ModuleI = (*Module)(nil)
 
+const LOGGER_MODULE = "ethereum.chain"
```
[the remark this answers is not in the document]
Comment (dongrie) [reply pkg/relay/ethereum/module.go]: Fix const name
https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/06014ce49818bb0e597d3e6c6e8c22d89b0923f9

File: pkg/relay/ethereum/module.go (hunk 1 above)
Comment (dongrie) [reply pkg/relay/ethereum/module.go:34]: Add `WithEthereumChain` to yui-relayer logger
https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/06014ce49818bb0e597d3e6c6e8c22d89b0923f9

**Comment by dongrie** [comment]
Remove WithEthereumChain logger. Use WithChain, WithChannel
https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/3323faed3a7d1d2cb109c0c8721c19bd3715ce80

File: pkg/relay/ethereum/tx.go [hunk 3]
```diff
@@ -262,3 +280,14 @@ func (c *Chain) TxAcknowledgement(opts *bind.TransactOpts, msg *chantypes.MsgAck
...
+func (c *Chain) GetChannelLogger() *log.RelayLogger {
+	logger := GetModuleLogger()
+	if c.Path() == nil {
+		return logger
+	}
+	chainID := c.Path().ChainID
+	portID := c.Path().PortID
+	channelID := c.Path().ChannelID
+	return logger.WithChannel(chainID, portID, channelID)
+}
```
Comment (siburu) [inline pkg/relay/ethereum/tx.go:293]: The logger created by `GetChannelLogger` is only suitable in the context of an established channel.
The following are examples of **NOT** suitable use cases.
- SendMsgs
    - This function is used for CreateClient or ConnectionOpenXxx or ChannelOpenXxx ...
- TxOpts
    - This function is used for sending all types of transactions, including ones inappropriate for `GetChannelLogger`

Comment (dongrie) [reply pkg/relay/ethereum/tx.go]: Only use channellogger for events

https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/260d6c02f22ed6a95b057e7df026522d351b59d6

Comment (siburu) [reply pkg/relay/ethereum/tx.go]: @dongrie thank you. please also use `GetChannelLogger` for `QueryUnreceivedXxx` and `QueryUnfinalizedXxx`.

Comment (dongrie) [reply pkg/relay/ethereum/tx.go]: https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/ec25e08df97694c8e6dc56ccb9fe9bbc751aad59
Fixed

File: pkg/relay/ethereum/tx.go [hunk 4]
```diff
@@ -101,8 +112,10 @@ func (c *Chain) TxCreateClient(opts *bind.TransactOpts, msg *clienttypes.MsgCrea
 }
 
 func (c *Chain) TxUpdateClient(opts *bind.TransactOpts, msg *clienttypes.MsgUpdateClient) (*gethtypes.Transaction, error) {
+	logger := c.GetChainLogger()
```
Comment (siburu) [inline pkg/relay/ethereum/tx.go:115]: Why did you only add logging to `TxCreateClient`, `TxUpdateClient`, `TxConnectionOpenTry` and `TxConnectionOpenAck`?
I think that protobuf marshaling basically never fails, and so logging here is unnecessary.

File: pkg/relay/ethereum/client.go [hunk 5]
```diff
@@ -23,6 +23,7 @@ func (chain *Chain) CallOpts(ctx context.Context, height int64) *bind.CallOpts {
 }
 
 func (chain *Chain) TxOpts(ctx context.Context) *bind.TransactOpts {
+	logger := chain.GetChainLogger()
```
Comment (siburu) [inline pkg/relay/ethereum/client.go:26]: Move this into the closure.

File: pkg/relay/ethereum/tx.go (hunk 4 above)
Comment (dongrie) [reply pkg/relay/ethereum/tx.go]: Remove logging 
https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/441f8e7d43fabe7e6f4dc54d2b20933d8408725c

File: pkg/relay/ethereum/client.go (hunk 5 above)
Comment (dongrie) [reply pkg/relay/ethereum/client.go]: Move to closure
https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/441f8e7d43fabe7e6f4dc54d2b20933d8408725c

File: pkg/relay/ethereum/events.go [hunk 6]
```diff
@@ -40,15 +42,20 @@ func init() {
...
 		chain.callOptsFromQueryContext(ctx),
 		chain.Path().PortID,
 		chain.Path().ChannelID,
 	); err != nil {
+		logger.Error("failed to get channel", err)
 		return nil, err
 	} else if !found {
-		return nil, fmt.Errorf("channel not found: sourcePortID=%v sourceChannel=%v", chain.Path().PortID, chain.Path().ChannelID)
+		err := fmt.Errorf("channel not found")
+		logger.Error("failed to get channel", err, "sourcePortID", chain.Path().PortID, "sourceChannel", chain.Path().ChannelID)
```
Comment (siburu) [inline pkg/relay/ethereum/events.go:57]: @dongrie thank you but change some keys.
- "sourcePortID" => "port_id"
- "sourceChannel" => "channel_id"

`chain` is not necessarily the source chain, so the original keys are confusing.

File: pkg/relay/ethereum/tx.go [hunk 7]
```diff
@@ -56,9 +59,11 @@ func (c *Chain) SendMsgs(msgs []sdk.Msg) ([]core.MsgID, error) {
...
 		// 	err = c.client.transfer(msg)
 		default:
+			logger.Error("failed to send msg", errors.New("illegal msg type"), "msg", msg)
 			panic("illegal msg type")
 		}
 		if err != nil {
+			logger.Error("failed to send msg", err, "msg", msg)
 			return nil, err
 		}
 		if receipt, revertReason, err := c.client.WaitForReceiptAndGet(ctx, tx.Hash(), c.config.EnableDebugTrace); err != nil {
```
Comment (siburu) [inline pkg/relay/ethereum/tx.go:69]: @dongrie please add logging to these error cases.

File: pkg/relay/ethereum/events.go (hunk 6 above)
Comment (dongrie) [reply pkg/relay/ethereum/events.go]: https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/4611488cb01d764157417b2ccbf882ec9e47d78f
Changed key name

File: pkg/relay/ethereum/tx.go (hunk 7 above)
Comment (dongrie) [reply pkg/relay/ethereum/tx.go:69]: Added `logger.Error`
https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/4611488cb01d764157417b2ccbf882ec9e47d78f

Comment (siburu) [reply pkg/relay/ethereum/tx.go:69]: @dongrie also add logging to the next else-if clause.

Comment (dongrie) [reply pkg/relay/ethereum/tx.go:69]: Sorry, I didn't check
I added `logger.Error` here as well
https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/2f9a74a137bf324cfb11b70b760d9b0313b96a4e

File: pkg/relay/ethereum/tx.go [hunk 8]
```diff
@@ -70,6 +70,7 @@ func (c *Chain) SendMsgs(msgs []sdk.Msg) ([]core.MsgID, error) {
 			logger.Error("failed to get receipt", err, "msg", msg)
 			return nil, err
 		} else if receipt.Status == gethtypes.ReceiptStatusFailed {
+			logger.Error("tx execution failed", fmt.Errorf("revertReason=%s, msgIndex=%d", revertReason, i), "msg", msg)
```
Comment (siburu) [inline pkg/relay/ethereum/tx.go:73]: I have a simple question. Why did you move "msg" from the error to the log attributes and leave "revertReason" and "msgIndex" in it?

Comment (dongrie) [reply pkg/relay/ethereum/tx.go:73]: https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/files/4611488cb01d764157417b2ccbf882ec9e47d78f..2f9a74a137bf324cfb11b70b760d9b0313b96a4e#diff-26026a3e1f9975f0f18cd6cec78a84c6a01ef33c3ee7d38af0bca97a08adb9ceL62

https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/files/4611488cb01d764157417b2ccbf882ec9e47d78f..2f9a74a137bf324cfb11b70b760d9b0313b96a4e#diff-26026a3e1f9975f0f18cd6cec78a84c6a01ef33c3ee7d38af0bca97a08adb9ceL66

This same pattern was used to output msg, revertReason, msgIndex to err.

Comment (siburu) [reply pkg/relay/ethereum/tx.go:73]: @dongrie In terms of structured logging, any useful values should be embedded as attributes.
On the other hand, it is also good to embed much information in errors.
Therefore, I think all attributes (revertReason, msg, msgIndex) should be embedded in both.
It would be a bit redundant, but seems most reasonable to me.
What do you think about it?

Comment (dongrie) [reply pkg/relay/ethereum/tx.go]: https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/2d359b6da1fd988b1ad4a65bd8e5e96009fe616d
msg is now included in err

File: pkg/relay/ethereum/tx.go [hunk 9]
```diff
@@ -56,19 +59,24 @@ func (c *Chain) SendMsgs(msgs []sdk.Msg) ([]core.MsgID, error) {
...
 			return nil, err
 		}
 		if receipt, revertReason, err := c.client.WaitForReceiptAndGet(ctx, tx.Hash(), c.config.EnableDebugTrace); err != nil {
+			logger.Error("failed to get receipt", err, "msg", msg)
 			return nil, err
 		} else if receipt.Status == gethtypes.ReceiptStatusFailed {
-			return nil, fmt.Errorf("tx execution failed: revertReason=%s, msgIndex=%d, msg=%v", revertReason, i, msg)
+			err := fmt.Errorf("tx execution failed: revertReason=%s, msgIndex=%d, msg=%v", revertReason, i, msg)
+			logger.Error("tx execution failed", err, "msg", msg)
+			return nil, err
```
Comment (siburu) [inline pkg/relay/ethereum/tx.go:75]: @dongrie Ah. Sorry, but my intention was/is the following...
```golang
err := fmt.Errorf("tx execution failed: revertReason=%s, msgIndex=%d, msg=%v", revertReason, i, msg)
logger.Error("tx execution failed", err, "revert_reason", revertReason, "msg_index", msgIndex, "msg", msg)
return nil, err
```

In my understanding, when structured logging is used, users of logs always obtain information from attributes  and never parse free-form error strings.
Therefore any important information (e.g. revertReason, msgIndex, msg) should be embedded as attributes.
My code above looks redundant but redundancy is better than lack of information.

Comment (dongrie) [reply pkg/relay/ethereum/tx.go]: https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/bcfa6f2e5e0b6e20c47c3a46398c25faee2c7b43
Change to use structured logging

**Review by siburu — APPROVED** [review]
